wit/window.wit

```diff
@@ -0,0 +1,228 @@
+        repeat: bool,
+    }
+    
+    flags mouse-buttons {
+        left,
+        middle,
+        right,
+    }
+    
+    record mouse {
+        position: position,
+        buttons: mouse-buttons,
+    }
+    
+    record scroll {
+        mouse: mouse,
+        x: f32,
+        y: f32,
+    }
+    
+    
+    /// Window system capabilities that are supported.
+    /// Methods that use unsupported capabilities will trap in WASM.
+    flags capabilities {
+        /// The windows have a title that can be set.
+        title,
+        /// The windows have position on a display which can be moved by the program. Not given e.g. on the web, where position is handled by CSS.
+        position,
+        // The windows have icons which can be set by the program.
+        icon,
+        /// Windows can request to go into fullscreen mode. This capability not being present can mean fullscreen windows
+        /// are not supported, or the window is forced fullscreen.
+        fullscreen,
+    }
+    
+    
+    record position {
+        x: u16,
+        y: u16,
+    }
+    
+    record size {
+        width: u16,
+        height: u16,
+    }
+    
+    record create-desc {
+        title: string,
+        position: position,
+        size: size,
+        visible: bool,
+    }
+    
+    /// Checks whether window functionality is available at all in the runtime.
+    available: func() -> bool;
+
+    resource window {
+        constructor(desc: create-desc);
+        
+        connect-graphics-context: func(context: borrow<graphics-context>);
+        
+        set-title: func(title: string);
+        get-title: func() -> string;
+        
+        // TODO add a display interface for querying the displays, so programs can see the resolution and find out where on which screen the window is.
+        set-position: func(position: position);
+        get-position: func() -> position;
+        
+        set-size: func(size: size);
+        get-size: func() -> size;
+        
+        /// Sets the visibility of the window.
+        set-visible: func(visible: bool);
+        
+        subscribe-resize: func() -> pollable;
+        get-resize: func() -> option<size>;
+        
+        subscribe-reposition: func() -> pollable;
+        get-reposition: func() -> option<position>;
+        
+        
+        subscribe-key-down: func() -> pollable;
+        get-key-down: func() -> option<key>;
+        
+        subscribe-key-up: func() -> pollable;
+        get-key-up: func() -> option<key>;
+        
+        
+        subscribe-mouse-down: func() -> pollable;
+        get-mouse-down: func() -> option<mouse>;
+        
+        subscribe-mouse-up: func() -> pollable;
+        get-mouse-up: func() -> option<mouse>;
+        
+        subscribe-mouse-move: func() -> pollable;
+        get-mouse-move: func() -> option<mouse>;
+        
+        subscribe-scroll: func() -> pollable;
+        get-scroll: func() -> option<scroll>;
```

Comment:
What does scrolling mean in this context? We don't really have document's that can be scrolled. Is this like [wheel events](https://developer.mozilla.org/en-US/docs/Web/API/Element/wheel_event) in JS?

Reply:
Yep, that's mouse wheel scroll events.

Reply:
If they only refer to the mouse wheel, would you mind naming it something like wheel? Since it won't necessarily be related to scrolling. E.g. some apps might use it for zooming, etc.

Reply:
https://github.com/WebAssembly/wasi-webgpu/pull/5/commits/5162b5e046c4a118121b7e7fac5da117bb517a76#diff-b4e12b94beccf6ff7685310d97d191fb8d72eca5028b87285ca5a08a5ea44059R222-R223